[CI] SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices

**Build scan**:
https://gradle-enterprise.elastic.co/s/hfll4qbiunmmy (7.x)
https://gradle-enterprise.elastic.co/s/7kqvtj64bdnew (master)
**Repro line**:

(from 7.x)
```
./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices" -Dtests.seed=26B49F246A6B424 -Dtests.locale=sr-Latn -Dtests.timezone=Africa/Bissau -Druntime.java=8
```

**Reproduces locally?**:
yes

**Applicable branches**:
7.x, master

**Failure history**:
Only failed [twice so far](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(_source),index:b646ed00-7efc-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:testUpgradeIndices),sort:!(process.time-start,desc))) but was recently added.

**Failure excerpt**:

org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests > testUpgradeIndices FAILED
    java.lang.AssertionError: 
    Expected: "frozen"
         but: was "normal"
        at __randomizedtesting.SeedInfo.seed([26B49F246A6B424:D64E14474C5532D1]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.lambda$testUpgradeIndices$0(SearchableSnapshotIndexMetadataUpgraderTests.java:70)
        at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1812)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449)
        at org.elasticsearch.xpack.searchablesnapshots.upgrade.SearchableSnapshotIndexMetadataUpgraderTests.testUpgradeIndices(SearchableSnapshotIndexMetadataUpgraderTests.java:58)

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgrader.java
@@ -92,7 +92,7 @@ public class SearchableSnapshotIndexMetadataUpgrader {
 
     static boolean needsUpgrade(ClusterState state) {
         return StreamSupport.stream(state.metadata().spliterator(), false)
-            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_7_13_0))
+            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0))
             .map(IndexMetadata::getSettings)
             .filter(SearchableSnapshotsConstants::isPartialSearchableSnapshotIndex)
             .anyMatch(SearchableSnapshotIndexMetadataUpgrader::notFrozenShardLimitGroup);
@@ -104,7 +104,7 @@ public class SearchableSnapshotIndexMetadataUpgrader {
         }
         Metadata.Builder builder = Metadata.builder(currentState.metadata());
         StreamSupport.stream(currentState.metadata().spliterator(), false)
-            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0) && imd.getCreationVersion().before(Version.V_7_13_0))
+            .filter(imd -> imd.getCreationVersion().onOrAfter(Version.V_7_12_0))
             .filter(
                 imd -> SearchableSnapshotsConstants.isPartialSearchableSnapshotIndex(imd.getSettings())
                     && notFrozenShardLimitGroup(imd.getSettings())

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgraderTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/upgrade/SearchableSnapshotIndexMetadataUpgraderTests.java
@@ -30,23 +30,37 @@ import static org.hamcrest.Matchers.sameInstance;
 public class SearchableSnapshotIndexMetadataUpgraderTests extends ESTestCase {
 
     public void testNoUpgradeNeeded() {
-        Metadata.Builder metadataBuilder = randomMetadata(normal(), full(), partial_7_13plus(), shardLimitGroupFrozen(partial_7_12()));
+        Metadata.Builder metadataBuilder = randomMetadata(
+            normal(),
+            full(),
+            shardLimitGroupFrozen(partial_7_13plus()),
+            shardLimitGroupFrozen(partialNeedsUpgrade())
+        );
         assertThat(needsUpgrade(metadataBuilder), is(false));
     }
 
     public void testNeedsUpgrade() {
-        Metadata.Builder metadataBuilder = addIndex(
-            partial_7_12(),
-            randomMetadata(normal(), full(), partial_7_13plus(), partial_7_12(), shardLimitGroupFrozen(partial_7_12()))
+        assertThat(
+            needsUpgrade(
+                addIndex(
+                    partialNeedsUpgrade(),
+                    randomMetadata(
+                        normal(),
+                        full(),
+                        partial_7_13plus(),
+                        partialNeedsUpgrade(),
+                        shardLimitGroupFrozen(partialNeedsUpgrade())
+                    )
+                )
+            ),
+            is(true)
         );
-        assertThat(needsUpgrade(metadataBuilder), is(true));
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/71973")
     public void testUpgradeIndices() {
         Metadata.Builder metadataBuilder = addIndex(
-            partial_7_12(),
-            randomMetadata(normal(), full(), partial_7_13plus(), partial_7_12(), shardLimitGroupFrozen(partial_7_12()))
+            partialNeedsUpgrade(),
+            randomMetadata(normal(), full(), partial_7_13plus(), partialNeedsUpgrade(), shardLimitGroupFrozen(partialNeedsUpgrade()))
         );
 
         ClusterState originalState = clusterState(metadataBuilder);
@@ -75,10 +89,17 @@ public class SearchableSnapshotIndexMetadataUpgraderTests extends ESTestCase {
                 return true;
             }
         }));
+
+        assertThat(SearchableSnapshotIndexMetadataUpgrader.needsUpgrade(upgradedState), is(false));
     }
 
     public void testNoopUpgrade() {
-        Metadata.Builder metadataBuilder = randomMetadata(normal(), full(), partial_7_13plus(), shardLimitGroupFrozen(partial_7_12()));
+        Metadata.Builder metadataBuilder = randomMetadata(
+            normal(),
+            full(),
+            partial_7_13plus(),
+            shardLimitGroupFrozen(partialNeedsUpgrade())
+        );
         ClusterState originalState = clusterState(metadataBuilder);
         ClusterState upgradedState = SearchableSnapshotIndexMetadataUpgrader.upgradeIndices(originalState);
         assertThat(upgradedState, sameInstance(originalState));
@@ -88,20 +109,21 @@ public class SearchableSnapshotIndexMetadataUpgraderTests extends ESTestCase {
         return settings(VersionUtils.randomVersion(random())).build();
     }
 
-    private Settings partial_7_12() {
-        return searchableSnapshotSettings(VersionUtils.randomVersionBetween(random(), Version.V_7_12_0, Version.V_7_12_1), true);
+    /**
+     * Simulate an index mounted with no shard limit group. Notice that due to not applying the group during rolling upgrades, we can see
+     * other than 7.12 versions here, but not 8.0 (since a rolling upgrade to 8.0 requires an upgrade to 7.latest first).
+     */
+    private Settings partialNeedsUpgrade() {
+        return searchableSnapshotSettings(VersionUtils.randomVersionBetween(random(), Version.V_7_12_0, Version.CURRENT), true);
     }
 
+    /**
+     * Simulate a 7.13plus mounted index with shard limit.
+     */
     private Settings partial_7_13plus() {
-        Settings settings = searchableSnapshotSettings(
-            VersionUtils.randomVersionBetween(random(), Version.V_7_13_0, Version.CURRENT),
-            true
+        return shardLimitGroupFrozen(
+            searchableSnapshotSettings(VersionUtils.randomVersionBetween(random(), Version.V_7_13_0, Version.CURRENT), true)
         );
-        if (randomBoolean()) {
-            return shardLimitGroupFrozen(settings);
-        } else {
-            return settings;
-        }
     }
 
     private Settings full() {

